app/models/dialog/orchestration_template_service_dialog.rb

```diff
@@ -0,0 +1,139 @@
+
+    def add_parameter_group(parameter_group, tab, position)
+      return if parameter_group.parameters.blank?
+
+      tab.dialog_groups.build(
+        :display  => "edit",
+        :label    => parameter_group.label || "Parameter Group#{position}",
+        :position => position
+      ).tap do |dialog_group|
+        parameter_group.parameters.each_with_index { |param, index| add_field(param, dialog_group, index, 'param_') }
+      end
+    end
+
+    def add_field(parameter, group, position, prefix = nil)
+      if parameter.constraints
+        dynamic_dropdown = parameter.constraints.detect { |c| c.kind_of?(OrchestrationTemplate::OrchestrationParameterAllowedDynamic) }
+        return create_dynamic_dropdown_list(parameter, group, position, dynamic_dropdown, prefix) if dynamic_dropdown
+
+        dropdown = parameter.constraints.detect { |c| c.kind_of?(OrchestrationTemplate::OrchestrationParameterAllowed) }
+        return create_dropdown_list(parameter, group, position, dropdown, prefix) if dropdown
+
+        checkbox = parameter.constraints.detect { |c| c.kind_of?(OrchestrationTemplate::OrchestrationParameterBoolean) }
+        return create_checkbox(parameter, group, position, prefix) if checkbox
+      end
+
+      create_textbox(parameter, group, position, prefix)
+    end
+
+    def create_dynamic_dropdown_list(parameter, group, position, dynamic_dropdown, prefix)
+      group.dialog_fields.build(
+        :type         => "DialogFieldDropDownList",
+        :name         => "#{prefix}#{parameter.name}",
+        :data_type    => parameter.data_type || "string",
+        :dynamic      => true,
+        :display      => "edit",
+        :required     => parameter.required,
+        :label        => parameter.label,
+        :description  => parameter.description,
+        :position     => position,
+        :dialog_group => group
+      ).tap do |dialog_field|
+        dialog_field.resource_action.fqname = dynamic_dropdown.fqname
+      end
+    end
+
+    def create_dropdown_list(parameter, group, position, dropdown, prefix)
+      values = dropdown.allowed_values
+      dropdown_list = values.kind_of?(Hash) ? values.to_a : values.collect { |v| [v, v] }
+      group.dialog_fields.build(
+        :type           => "DialogFieldDropDownList",
+        :name           => "#{prefix}#{parameter.name}",
+        :data_type      => parameter.data_type || "string",
+        :display        => "edit",
+        :required       => parameter.required,
+        :values         => dropdown_list,
+        :default_value  => parameter.default_value || dropdown_list.first,
+        :label          => parameter.label,
+        :description    => parameter.description,
+        :reconfigurable => parameter.reconfigurable,
+        :position       => position,
+        :dialog_group   => group
+      )
+    end
+
+    def create_textbox(parameter, group, position, prefix)
+      if parameter.data_type == 'string' || parameter.data_type == 'integer'
```

Comment:
What are the other field types we could expect to see here?  From the previous logic it is only `json`.  Since we are overriding the `data_type` in the `else` case it would it make more sense to pass a `data_type` value that would defined that we want a TextAreaBox?

Example: `parameter.data_type = string_text_box`

Reply:
When the execution calls into this method we know a text box is needed. It is a matter of deciding single line or multiple lines. I consider `string` and `integer` that need a single line. Everything else uses multiple lines to be safe. With this relaxation we can accept user defined data types.